### DISCUSSION TITLE
Native osu! modcharts?

### DISCUSSION THREAD

**Discussion by plaguemask** [ppy/osu#16516] [Ideas]
Thanks to a [mesmerizing showcase at AGDQ](https://www.youtube.com/watch?v=sBP8MxQhEVM), I've recently been introduced to the world of StepMania modcharts, in which creators add a multitude of visual effects in order to push the boundaries of how well players can read confusing and distracting charts.

This has made me wonder about the possibility of something like this in osu!. I play with Fun mods often, but mods can be turned off and maps weren't designed with them in mind. A lot is possible with storyboards, like the Aspire 2018 winner, but maps like those only work by way of quirks/bugs in stable. In my mind, "real" support would include the ability to visually alter hit objects (and maybe certain UI elements) in a way that's inherent to the map.

How possible would it even be to include this in osu! itself? How about in a custom ruleset, or in osu!framework at all? I just think the idea is neat.

**Comment by bdach** [comment]
there have been instances of doing this via storyboards, but the unfortunate thing is that a lot of the capabilities required to make these work also inadvertently double for things that can be potentially exploited to damage the integrity of competitive gameplay. as an example, it turned out that people could skin osu!taiko in a way that allowed them to display the playfield as if they were playing mania, and the fix for that exploit ended up inadvertently impacting some existing taiko modmaps.

see: https://github.com/ppy/osu-stable-issues/issues/391, https://github.com/ppy/osu-stable-issues/issues/651

**Reply by plaguemask** [reply]
I see how that can be a problem with skinning, but I'm more imagining a system that's contained to individual beatmaps. Like an enhanced storyboard that can animate/distort hit objects and the play field — no map skin needed. Whether that is even feasible though is really the heart of my question.

**Reply by bdach** [reply]
feasible? maybe. happening anytime soon? probably not. difficult to put a timeline on this even, may be years before it happens if it even does. many other things to work on first that have higher priority.

**Reply by plaguemask** [reply]
Understandable! That's about the answer I was expecting, but I wanted to float the idea anyway just to be sure, and because the possibilities are quite inspiring.
